### PR TITLE
fix(`PromiseReleaseReconciler`): Fail if installing `Promise` fails

### LINKED ISSUE #87: bug: when the Promise is invalid, installing it via Promise release gives you no feedback
### Steps to reproduce
1. Create an invalid promise
1. Add the following label to it: `kratix.io/promise-version: v1.0.0+beta.1`
2. Try to install this promise
3. See that it fails
4. Create a Promise Release pointing to this invalid promise
```yaml
apiVersion: platform.kratix.io/v1alpha1
kind: PromiseRelease
metadata:
  name: promise
spec:
  version: v1.0.0+beta.1
  sourceRef:
    type: http
    url: <url>
```
5. Apply the Promise release

### Expected

Some feedback

### Actual

None

### PR DESCRIPTION
**Pull request by Bhargav-InfraCloud**
## Description
Fail the `PromiseReleaseReconciler` if there is an error installing the referenced promise.

Add an event to PromiseRelease to state the reason for the failure of installing Promise. If it fails because of an invalid Promise manifest referenced, state the reason in the event accordingly.

Add RBAC permissions to `PromiseReleaseReconciler` to `CREATE` and `PATCH` `events` on itself in-cluster.

Fixes #87

### 1. Create clusters and install Kratix.
```
make quick-start
```
### 2. Create `PromiseRelease` with a reference URL to an invalid `Promise` (eg. incorrect label value syntax).
```
git show issue-87-temp:promiserelease.yaml | kubectl apply -f -
```
> Note:  [issue-87-temp][1] is a temporary branch holding a `PromiseRelease` file [here][2], and more importantly, the invalid `Promise` file [here][3]. This is directly applied from the branch and remote. If the branch is missing locally, do a `git fetch` before running the above command.
### 3. Check `Promise` in the cluster.  `Note`: It doesn't get created due to an invalid label, with or without the current changes.
```
kubectl get promises.platform.kratix.io
```
> Note: `Promise` is cluster-scoped.
### 4. Check `PromiseRelease`'s status.
```
kubectl get promisereleases.platform.kratix.io/promise
```
> Note: `PromiseRelease` is cluster-scoped.
#### Earlier
```
NAME      STATUS      VERSION
promise   Installed   v1.0.0+beta.1
```
#### Now
```
NAME      STATUS             VERSION
promise   Error installing   v1.0.0+beta.1
```
### 5. Check `PromiseRelease`'s events.
```
kubectl describe promisereleases.platform.kratix.io/promise
```
#### Earlier
```
Events:                    <none>
```
#### Now
```
Events:
  Type     Reason           Age                From                      Message
  ----     ------           ----               ----                      -------
  Warning  Invalid Promise  1s (x14 over 42s)  PromiseReleaseController  Failed to install Promise "nginx-ingress": Promise.platform.kratix.io "nginx-ingress" is invalid: metadata.labels: Invalid value: "v1.0.0+beta.1": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')
```

[1]: https://github.com/Bhargav-InfraCloud/kratix-ic/tree/issue-87-temp
[2]: https://github.com/Bhargav-InfraCloud/kratix-ic/blob/issue-87-temp/promiserelease.yaml
[3]: https://github.com/Bhargav-InfraCloud/kratix-ic/blob/issue-87-temp/invalid-promise.yaml

### CODE REVIEW

**Comment by Bhargav-InfraCloud** [comment]
CC: kirederik

**Review by kirederik — CHANGES_REQUESTED** [review]

File: controllers/promiserelease_controller_test.go [hunk 1]
```diff
@@ -402,4 +405,89 @@ var _ = Describe("PromiseReleaseController", func() {
...
+						err := fakeK8sClient.Create(context.TODO(), &promiseRelease)
+						Expect(err).NotTo(HaveOccurred())
+					})
+
+					It("fails the PromiseRelease's reconciliation", func() {
+						_, err := t.reconcileUntilCompletion(reconciler, &promiseRelease)
+						Expect(err).To(HaveOccurred())
+						Expect(<-eventRecorder.Events).To(ContainSubstring(
+							fmt.Sprintf("Promise.platform.kratix.io %q is invalid", promise.Name)),
+						)
```
[the remark this answers is not in the document]
Comment (Bhargav-InfraCloud) [reply controllers/promiserelease_controller_test.go]: Ah! Didn't realize there's a func in Gomega for the same. Thanks! I updated it.

File: controllers/promiserelease_controller_test.go [hunk 2]
```diff
@@ -402,4 +405,89 @@ var _ = Describe("PromiseReleaseController", func() {
 			Expect(promise.Spec).To(Equal(expectedPromise.Spec))
 		})
 	})
+
+	// This test is separated from the original nested tests in order to test using mocked kubeclient's method(s)
```
[the remark this answers is not in the document]
Comment (Bhargav-InfraCloud) [reply controllers/promiserelease_controller_test.go]: You are right. Overwriting it should do. Done. Thanks!

**Review by kirederik — APPROVED** [review]
good stuff! Thanks!

We'll sort the CLA and get this one merged soon!

**Comment by Bhargav-InfraCloud** [comment]
Signed the CLA. Thanks!

**Comment by kirederik** [comment]
Awesome! Thanks @Bhargav-InfraCloud and thanks for your patience.

**Comment by Bhargav-InfraCloud** [comment]
Thanks @kirederik! 😊

**Pull request merged**